### ISSUE TITLE
date: wrong timezone for 1970 -> 1971 in Europe/London timezone

### ISSUE THREAD

**Issue by stephane-chazelas** [ksh93/ksh#669]
Cloned from https://github.com/att/ast/issues/17 as it still applies to

```
  version         sh (AT&T Research) 93u+m/1.1.0-alpha+3c2b0dcf 2023-06-14
```

https://en.wikipedia.org/wiki/British_Summer_Time#Periods_of_deviation

Between 27 October 1968 and 31 October 1971, there was no daylight saving time in mainland Britain. It was GMT+1 all year round. On both GNU and Solaris system, the system's strftime/localtime is correct:

```
$ TZ=Europe/London perl -MPOSIX -le 'print strftime "%F %T %Z %z", localtime 0'
1970-01-01 01:00:00 BST +0100
```

(BST being then British _Standard_ Time, not _Summer_ this time).

But ksh93's "printf %T" or the ast date utility seem to get it wrong:

```
$ ksh93 -c 'printf "%(%F %T %Z %z)T\n" "#0"'
1970-01-01 00:00:00 GMT -0000
```

(ksh93u on Solaris 10)

```
$ arch/linux.i386-64/bin/date -d "#0"
Thu Jan  1 00:00:00 GMT 1970
```

(from the beta branch on Debian).

**Comment by pghvlaans** [comment]
I've done a little testing, and it appears that anachronistic modern time zone labels are being used.

Today, no offset in London means `GMT`, so that's what shows up in the 1970-1971 period when there is no offset (rather than `BST`), and the time is one hour behind what it should be. If the idea of "no offset" for `BST` in January sounds strange, consider the `tz` Europe data table:

    # Summer Time Order, 1968 (S.I. 1968/117)
    Rule	GB-Eire	1968	only	-	Feb	18	2:00s	1:00	BST
    # The British Standard Time Act, 1968
    #	(no summer time)
    # The Summer Time Act, 1972
    Rule	GB-Eire	1972	1980	-	Mar	Sun>=16	2:00s	1:00	BST

No offset existed at all for the then-active GB-Eire rule at that time period, so none is detected.

This is not restricted to Europe/London. Just as an example, using epoch 236961303 (1977-07-05), similar patterns appear for Europe/Riga (today `EET`/`EEST`, then `MSK`) and Europe/Istanbul (today `03`, then `EET`/`EEST`):

    $ TZ=Europe/Riga printf "%(%F %T %Z %z)T\n" "#236961303"   
    1977-07-05 16:35:03 EET +0200
    $ TZ=Europe/Riga date --date="@236961303" "+%Y-%m-%d %H:%M:%S %Z %z"
    1977-07-05 17:35:03 MSK +0300
    
    $ TZ=Europe/Istanbul printf "%(%F %T %Z %z)T\n" "#236961303"
    1977-07-05 20:35:03 +03 +0600
    $ TZ=Europe/Istanbul date --date="@236961303" "+%Y-%m-%d %H:%M:%S %Z %z"
    1977-07-05 17:35:03 EEST +0300

**Comment by pghvlaans** [comment]
~~I believe I have a working solution for `printf`, but I haven't done any testing on `date`. May I ask how you built it? `ALL_LIBCMD=1` didn't work for me.~~

Ah, never mind, I spaced and didn't realize the last example was from ksh93u. Sorry about that.

Anyway, I'll have a PR ready to go once my test runs have wrapped up.

**Comment by pghvlaans** [comment]
Spoke too soon. Issues on Void (musl) and Haiku.

**Comment by McDutchie** [comment]
Fixed in 98800e9e18634cabac8046f6e9f7affec85b3d51 (dev), 580bc2d954b0210de2a6ab3088d3c396f86b0c09 (1.0)
